src/wallet/rpcwallet.cpp

```diff
@@ -702,19 +711,84 @@ UniValue getbalance(const UniValue& params, bool fHelp)
 
             int64_t allFee;
             string strSentAccount;
-            list<pair<CTxDestination, int64_t> > listReceived;
-            list<pair<CTxDestination, int64_t> > listSent;
+            list<COutputEntry> listReceived;
+            list<COutputEntry> listSent;
             wtx.GetAmounts(listReceived, listSent, allFee, strSentAccount, filter);
             if (wtx.GetDepthInMainChain() >= nMinDepth && wtx.GetBlocksToMaturity() == 0)
             {
                 for (auto const& r : listReceived)
-                    nBalance += r.second;
+                {
+                    CBitcoinAddress addr;
+                    addr.Set(r.destination);
+
+                    nBalance += r.amount;
+
+                    if (detailed_output)
+                    {
+                        UniValue item(UniValue::VOBJ);
+
+                        item.pushKV("timestamp", (int64_t) wtx.nTime);
+                        item.pushKV("hash", wtx.GetHash().ToString());
+                        item.pushKV("address", addr.ToString());
+                        item.pushKV("amount", ValueFromAmount(r.amount));
+
+                        items.push_back(item);
+                    }
+                }
             }
-            for (auto const& r : listSent)
-                nBalance -= r.second;
+            for (auto const& s : listSent)
+            {
+                CBitcoinAddress addr;
+                addr.Set(s.destination);
+
+                nBalance -= s.amount;
+
+                if (detailed_output)
+                {
+                    UniValue item(UniValue::VOBJ);
+
+                    item.pushKV("timestamp", (int64_t) wtx.nTime);
+                    item.pushKV("hash", wtx.GetHash().ToString());
+                    item.pushKV("address", addr.ToString());
+                    item.pushKV("amount", ValueFromAmount(-s.amount));
+
+                    items.push_back(item);
+                }
+            }
+
             nBalance -= allFee;
+
+            if (detailed_output)
+            {
+                totalFee += allFee;
+
+                UniValue item(UniValue::VOBJ);
+
+                if (allFee)
+                {
+                    item.pushKV("timestamp", (int64_t) wtx.nTime);
+                    item.pushKV("hash", wtx.GetHash().ToString());
+                    item.pushKV("address", std::string {});
+                    item.pushKV("fee", ValueFromAmount(allFee));
+
+                    items.push_back(item);
+                }
+
+            }
+        }
+
+        if (detailed_output)
+        {
+            ret.pushKV("balance", ValueFromAmount(nBalance));
+            ret.pushKV("fees", ValueFromAmount(totalFee));
+            ret.pushKV("list", items);
+
+            return ret;
+        }
+        else
+        {
+            return ValueFromAmount(nBalance);
```

Comment:
> What I discovered is that the existing dev branch (WITHOUT THIS PR) gives wildly inaccurate results for the following two balance retrieval methods:
> 
> getbalance: 219249.21897026 (agrees with the GUI)
> getbalance "*" 0: 81234.63440220
> 
> These two different methods should return the same balance. The first basically adds up all unspent outputs in the wallet, the second goes through each and every transaction in the wallet and does a net total.

We're going to remove the deprecated accounting API completely in the release after Fern. Bitcoin adds `CWallet::GetBalance(const int min_depth, bool avoid_reuse)` and just forwards to this call from `getbalance` whether or not the account argument is set. I think it's worth future-proofing this now by enforcing the accounting API deprecation in `getbalance` and adding this overload to centralize the logic rather than trying to twist the code around an option that won't exist.

Reply:
Then maybe it really isn't an issue. The balance is very close now, and inspection by hand shows that listtransactions and gettransaction is producing the right output IMHO...

Reply:
Do you think any of the exchanges are using the accounting API anyway?

Reply:
One other comment... when you put "*" as the "account" argument to getbalance, it actually bypasses the accounting calls and just uses ListTransactions and GetAmounts...

Reply:
I doubt it. I imagine that exchanges are well aware of the deprecation. A user needs to set `-enableaccounts` to even get access to all of the old API. If I built a service against the RPC, I'd wouldn't touch something like that with a ten foot pole. For something like an exchange, it's more flexible and efficient to manage accounts via a proper database than through the wallet anyway.

Reply:
So I think we are pretty safe. I am removing the draft designator from this PR.